ulog.cpp

```diff
@@ -0,0 +1,415 @@
+            memset(((uint8_t *)&cmd) + payload_len, 0, trimmed_zeros);
+
+        if (!_timeout || cmd.command != MAV_CMD_LOGGING_START)
+            break;
+
+        if (cmd.result == MAV_RESULT_ACCEPTED) {
+            _mainloop->del_timeout(_timeout);
+            _timeout = NULL;
+        } else
+            log_error("MAV_CMD_LOGGING_START result(%u) is different than accepted", cmd.result);
+        break;
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA_ACKED: {
+        mavlink_logging_data_acked_t *ulog_data_acked = (mavlink_logging_data_acked_t *)payload;
+        mavlink_message_t msg;
+        mavlink_logging_ack_t ack;
+
+        ack.sequence = ulog_data_acked->sequence;
+        ack.target_component = MAV_COMP_ID_ALL;
+        ack.target_system = _target_system_id;
+        mavlink_msg_logging_ack_encode(SYSTEM_ID, MAV_COMP_ID_ALL, &msg, &ack);
+        _send_msg(&msg, _target_system_id);
+        /* no break needed, message will be handled by MAVLINK_MSG_ID_LOGGING_DATA case */
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA: {
+        if (trimmed_zeros) {
+            mavlink_logging_data_t ulog_data;
+            memcpy(&ulog_data, payload, payload_len);
+            memset(((uint8_t *)&ulog_data) + payload_len, 0, trimmed_zeros);
+            _logging_data_process(&ulog_data);
+        } else {
+            mavlink_logging_data_t *ulog_data = (mavlink_logging_data_t *)payload;
+            _logging_data_process(ulog_data);
+        }
+        break;
+    }
+    }
+
+    return buffer->len;
+}
+
+/*
+ * Return true if the message with seq should be handled.
+ */
+bool ULog::_logging_seq(uint16_t seq, bool *drop)
+{
+    if (_expected_seq == seq) {
+        _expected_seq++;
+        *drop = false;
+        return true;
+    }
+
+    if (seq > _expected_seq) {
+        const uint16_t diff = seq - _expected_seq;
+        if (diff > (UINT16_MAX / 2)) {
+            /* _expected_seq wrapped and a re-transmission of a non-wrapped message happened */
+            return false;
+        }
+    } else {
+        const uint16_t diff = _expected_seq - seq;
+        if (diff < (UINT16_MAX / 2)) {
+            /* re-transmission */
+            return false;
+        }
+    }
+
+    *drop = true;
+    _expected_seq = seq + 1;
+    return true;
+}
+
+void ULog::_logging_data_process(mavlink_logging_data_t *msg)
+{
+    bool drops = false;
+
+    if (!_logging_seq(msg->sequence, &drops))
+        return;
+
+    /* Waiting for ULog header? */
+    if (_waiting_header) {
+        const uint8_t magic[] = ULOG_MAGIC;
+
+        if (memcmp(magic, msg->data, sizeof(magic))) {
+            log_error("Invalid ULog Magic number");
+            return;
+        }
+        if (msg->length < ULOG_HEADER_SIZE) {
+            /* This should never happen */
+            log_error("ULog header is not completed");
+            return;
+        }
+
+        const ssize_t r = write(_file, msg->data, ULOG_HEADER_SIZE);
+        if (r != ULOG_HEADER_SIZE) {
```

Comment:
Is this check meaningful? I mean, `_file` is `O_NONBLOCK`, so does it return anything meaningful here? AFAICU, it all depends on underlying implementation - default being just ignoring `O_NONBLOCK` for regular files. Default makes check meaningful - but `O_NONBLOCK` not.